Storage Add Ons: Move UI state into wpcom-plans-ui data store

jeyip opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Follow-up to https://github.com/Automattic/wp-calypso/pull/79041#discussion_r1277445769

> Just want to reemphasize this a little. Our immediate next priority for the plans package is https://github.com/Automattic/wp-calypso/issues/78266, which is about breaking apart the grid components and exporting individually. This is my next issue on the queue.

> So we need to keep our focus/priorities on the refactor we discussed previously on a few occassions (to remove the local `selectedStorage` state from `PlanFeatures2023Grid` and use the `wpcom-plan-ui` data store instead)

## Proposed Changes

[As part of the plans 2023 pricing page refactor](https://github.com/Automattic/wp-calypso/issues/78266), @chriskmnds is breaking apart the pricing grid components ( features grid & comparison grid ) and exporting them individually. 

To prevent roadblocks to his work, **we want to avoid sharing UI state locally between the features grid and the comparison grid**. As a result, we're moving any storage add ons UI state away from local state management and into the `wpcom-plan-ui` data store.

## GIF
![2023-08-02 17 37 45](https://github.com/Automattic/wp-calypso/assets/5414230/b82d131f-d50d-470b-bf82-825ff4130e30)

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Spin up a local dev environment or use calypso live
* Navigate to `/start/plans?flags=plans/upgradeable-storage`
* Verify that selecting items in the storage add ons dropdown continues to behave as expected. To be clear, all we're checking is that dropdown selections persist. No other functionality is wired up at this point
* Do the same for /plans

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

jeyip committed:
Move storage add on UI state into wpcom-plans-ui

jeyip committed:
Add calypso products dependency

jeyip committed:
Update yarn lock file

jeyip commented on packages/data-stores/src/wpcom-plans-ui/reducer.ts:
```diff
@@ -14,8 +15,19 @@ const showDomainUpsellDialog: Reducer< boolean | undefined, WpcomPlansUIAction >
 	return state;
 };
 
+const selectedStorageAddOnsForPlans: Reducer<
```
**Note:**

Naming this slice of state felt odd because I wasn't sure how to interpret the existing convention. A sibling reducer is called [showDomainUpsellDialog](https://github.com/Automattic/wp-calypso/pull/80158/files#diff-e00c92dec12db0d10ab243c2c8b28430e8800a62f4bd3c844589c07ad7630e21R6). In the end, I tried to make the name as specific as possible, but I'm totally open to changing this

jeyip commented:
@chriskmnds would you mind taking a look at this?

chriskmnds commented on packages/data-stores/src/wpcom-plans-ui/actions.ts:
```diff
@@ -9,4 +9,19 @@ export const resetStore = () =>
+	plan: string;
+} ) =>
+	( {
+		type: 'WPCOM_PLANS_UI_SET_STORAGE_ADD_ON_FOR_PLAN',
```
Thinking we can probably include "selected" in these selectors/constants/actions. Maybe also replace "addOn" with "option" (or both), as a little more obvious it's an option from a list? So `setSelectedStorageOptionForPlan`, etc. Of course, merely a suggestion, as you see fit.

jeyip replied:
I initially went with the term add on where possible because it matches the vocabulary we're using for the UX, product, and business side of things, _but_ since this is the data store for plans UI, I think you're right.

Mulling this over, but will probably change the name 🤔

jeyip replied:
https://github.com/Automattic/wp-calypso/pull/80158/commits/d95e8c08f6d2b0989be7ec4333c7dfded49ede93

chriskmnds commented on packages/data-stores/src/wpcom-plans-ui/actions.ts:
```diff
@@ -9,4 +9,19 @@ export const resetStore = () =>
+	plan,
+}: {
+	addOnSlug: string;
+	plan: string;
```
I think let's rename `plan` to `planSlug` here. Does it help to generalize into "plan"? 

Also, the type can be `PlanSlug` (or was there anything blocking from doing so?)

jeyip replied:
> I think let's rename plan to planSlug here. Does it help to generalize into "plan"?

Taking a look at this with fresh eyes after being away for a bit ( working on the component migrations ). I agree with using planSlug instead 👍 

> Also, the type can be PlanSlug (or was there anything blocking from doing so?)

I've forgotten if there was a reason, but I'll make some updates and see

jeyip replied:
https://github.com/Automattic/wp-calypso/pull/80158/commits/6aeef7cb587d2ec7aaa5286999f6e0db5eb4afc7

chriskmnds commented on packages/data-stores/package.json:
```diff
@@ -54,6 +54,7 @@
 		"react-dom": "^18.2.0"
 	},
 	"devDependencies": {
+		"@automattic/calypso-products": "workspace:^",
```
I believe we need to release `@automattic/calypso-products` to NPM before including it as a dependency to data-stores package (that is NPM-released). It should be a [simple background process](https://github.com/Automattic/wp-calypso/blob/trunk/docs/monorepo.md#publishing) for the most part, but may need to audit and add a Changelog to the package first (and an "initial release" entry there). 

Feel free to ask around as I'm not too familiar with how `workspace` dependencies work, but considering always that the NPM data-stores package can be built outside of Calypso (into a Gutenberg component, for example).

jeyip replied:
> I believe we need to release @automattic/calypso-products to NPM before including it as a dependency to data-stores package (that is NPM-released). It should be a [simple background process](https://github.com/Automattic/wp-calypso/blob/trunk/docs/monorepo.md#publishing) for the most part, but may need to audit and add a Changelog to the package first (and an "initial release" entry there).

Woah this is a great catch 🎉  Thanks for noting this. I had assumed it was published already. I'll double check and figure out how to release it if necessary

_Edit:_
Confirming here p1691705544283559-slack-C7YPUHBB2

jeyip replied:
Started the process by creating an initial changelog here https://github.com/Automattic/wp-calypso/pull/80474

jeyip replied:
https://github.com/Automattic/wp-calypso/pull/80474 is merged and it looks like a special audit may not be required p1691761891545059/1691731751.646959-slack-C03N25JPCE4.

I ended up manually spot checking for any passwords or keys but didn't see anything. @automattic/calypso-products, is now published https://www.npmjs.com/package/@automattic/calypso-products

chriskmnds commented on packages/data-stores/src/wpcom-plans-ui/reducer.ts:
```diff
@@ -14,8 +15,19 @@ const showDomainUpsellDialog: Reducer< boolean | undefined, WpcomPlansUIAction >
+
 const reducer = combineReducers( {
 	showDomainUpsellDialog,
+	selectedStorageAddOnsForPlans,
```
Asking here, so we can resolve in a thread (so not exactly related to this piece of code).

Do we expect to retrieve the selected storage from a location outside of the plans grid package (`plan-features-2023-grid`)? e.g. from `checkout` maybe? If that's the case, then we will need to set the value to `persist` in the data store, and possibly call an action to reset it when rendering the grid components (features table or comparison grid) in a "fire once" hook.

jeyip replied:
> Do we expect to retrieve the selected storage from a location outside of the plans grid package (plan-features-2023-grid)? e.g. from checkout maybe?

Not for the time being. `checkout` already has their own state management in place for add ons

chriskmnds reviewed (changes requested):
Thanks for working on this! Left some comments below, https://github.com/Automattic/wp-calypso/pull/80158#discussion_r1286956303 seems critical and merits investigation. Overall looks good to me though.

Sharpirate commented:
#### Testing
✔️ The dropdown works as expected both with keyboard and mouse. The selection is persisted.
✔️ The payload of the corresponding Redux actions looks good:

• 50 GB:
```
type:"WPCOM_PLANS_UI_SET_STORAGE_ADD_ON_FOR_PLAN"
addOnSlug:"50gb-storage"
plan:"business-bundle"
```

• 100 GB:
```
type:"WPCOM_PLANS_UI_SET_STORAGE_ADD_ON_FOR_PLAN"
addOnSlug:"50gb-storage-add-on"
plan:"business-bundle"
```
• 150 GB:
```
type:"WPCOM_PLANS_UI_SET_STORAGE_ADD_ON_FOR_PLAN"
addOnSlug:"100gb-storage-add-on"
plan:"business-bundle"
```

sharpirate commented on packages/data-stores/src/wpcom-plans-ui/selectors.ts:
```diff
@@ -1,3 +1,7 @@
 import type { State } from './reducer';
 
 export const isDomainUpsellDialogShown = ( state: State ) => !! state.showDomainUpsellDialog;
+export const getStorageAddOnForPlan = ( state: State ) => ( plan: string ) => {
```
Just out of curiosity because I haven't seen this before, why do we need a function returning a function? Can't we supply the `plan` to `getStorageAddOnForPlan` directly?

chriskmnds replied:
> Can't we supply the plan to getStorageAddOnForPlan directly?

Thank you @Sharpirate . Good catch, and yes, we should. No need for currying here.

@jeyip

jeyip replied:
Thanks y'all! Resolved in https://github.com/Automattic/wp-calypso/pull/80158/commits/072842751fed173b43d931f19fa69fb849deb274

Sharpirate commented:
> Do the same for /plans
I wasn't able to test `/plans` because when I go to `http://calypso.localhost:3000/plans` I get redirected to `https://wordpress.com/pricing/`, not sure if I'm missing something

jeyip commented:
Thanks for all the feedback y'all! Taking a look now 👍

jeyip committed:
Use setSelectedStorageOptionForPlan as action name

jeyip committed:
Update parameter names and types in setSelectedStorageOptionForPlan

jeyip committed:
Fix typescript errors

jeyip committed:
Fix code styling error

chriskmnds commented on packages/data-stores/src/wpcom-plans-ui/selectors.ts:
```diff
@@ -1,3 +1,7 @@
 import type { State } from './reducer';
 
 export const isDomainUpsellDialogShown = ( state: State ) => !! state.showDomainUpsellDialog;
+export const getStorageAddOnForPlan = ( state: State ) => ( plan: string ) => {
```
`plan` -> `planSlug` applies for here too. Aso a proper type.

jeyip replied:
Good point -- made the change in [2b80eda](https://github.com/Automattic/wp-calypso/pull/80158/commits/2b80eda51d7c6308773462ee5b0a472f28b845aa)

chriskmnds commented on packages/data-stores/src/wpcom-plans-ui/selectors.ts:
```diff
@@ -1,3 +1,7 @@
 
 export const isDomainUpsellDialogShown = ( state: State ) => !! state.showDomainUpsellDialog;
+export const getStorageAddOnForPlan = ( state: State ) => ( plan: string ) => {
+	const selectedStorageAddOnsForPlans = state?.selectedStorageAddOnsForPlans;
```
Can `state` be undefined?
and is there a need for the intermediate variable/assignment?

jeyip replied:
Ah also a good catch. It's selectedStorageAddOnsForPlans that initializes as `undefined`, not `state` itself.

I originally added it for readability, but no strong opinions. I'll remove it since it seems like it might not be helping here though 👍 

I also switched from returning `null` to `undefined` if a storageAddOn isn't found since it seems more appropriate when a value can't be located

https://github.com/Automattic/wp-calypso/pull/80158/commits/87d4c887aa04c2bd8a9a29470302011d04b6114c

jeyip committed:
Update plan slug type

jeyip committed:
Fix optional chaining for getStorageAddOnForPlan selector

State doesn't initialize as undefined -- it's the substate selectedStorageAddOnsForPlans that does. We update the optional chaining of the getStorageAddOnForPlan selector to reflect this

jeyip committed:
Remove unnecessary currying

jeyip committed:
Use calypso-products storageOption type

jeyip committed:
Remove null value type

sharpirate reviewed (approved):
Code looks good and the PR tests well after the recent changes. 👍

jeyip committed:
Remove unnecessary plan features grid state type

jeyip commented:
To avoid any more merge conflicts with this PR I'll be merging it shortly.

All comments have been addressed, but @chriskmnds if there's any more you'd like to add when you're back from AFK, I'll be happy to draft immediate follow-ups for any new concerns that pop up.
